dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/ListTelnetHandler.java

```diff
@@ -46,55 +51,107 @@ public String telnet(Channel channel, String message) {
                     detail = true;
                 } else {
                     if (service != null && service.length() > 0) {
-                        return "Invaild parameter " + part;
+                        return "Invalid parameter " + part;
                     }
                     service = part;
                 }
             }
         } else {
             service = (String) channel.getAttribute(ChangeTelnetHandler.SERVICE_KEY);
             if (service != null && service.length() > 0) {
-                buf.append("Use default service " + service + ".\r\n");
+                buf.append("Use default service ").append(service).append(".\r\n");
             }
         }
+
         if (service == null || service.length() == 0) {
-            for (Exporter<?> exporter : DubboProtocol.getDubboProtocol().getExporters()) {
-                if (buf.length() > 0) {
-                    buf.append("\r\n");
-                }
-                buf.append(exporter.getInvoker().getInterface().getName());
-                if (detail) {
-                    buf.append(" -> ");
-                    buf.append(exporter.getInvoker().getUrl());
-                }
-            }
+            printAllServices(buf, detail);
         } else {
-            Invoker<?> invoker = null;
-            for (Exporter<?> exporter : DubboProtocol.getDubboProtocol().getExporters()) {
-                if (service.equals(exporter.getInvoker().getInterface().getSimpleName())
-                        || service.equals(exporter.getInvoker().getInterface().getName())
-                        || service.equals(exporter.getInvoker().getUrl().getPath())) {
-                    invoker = exporter.getInvoker();
-                    break;
+            printSpecifiedService(service, buf, detail);
+
+            if (buf.length() == 0) {
+                buf.append("No such service: ").append(service);
+            }
+        }
+        return buf.toString();
+    }
+
+    private void printAllServices(StringBuilder buf, boolean detail) {
+        printAllProvidedServices(buf, detail);
+        printAllReferredServices(buf, detail);
+    }
+
+    private void printAllProvidedServices(StringBuilder buf, boolean detail) {
```

Comment:
Can be simplified to 
`private void printAllProvidedServices(StringBuilder buf, boolean detail) {
        if(!ApplicationModel.allProviderModels().isEmpty()) {
            buf.append("PROVIDER:\r\n");
        }
        for (ProviderModel provider : ApplicationModel.allProviderModels()) {
            buf.append(provider.getServiceName());
            if (detail) {
                buf.append(" -> ");
                buf.append(" published: ");
                buf.append(isRegistered(provider.getServiceName()) ? "Y" : "N");
            }
            buf.append("\r\n");
        }
    }`

Reply:
done.